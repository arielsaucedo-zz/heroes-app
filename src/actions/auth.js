import axios from "axios";

import { types } from "../types/types";

/* Accion: login de usuario */
export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await axios.post("http://challenge-react.alkemy.org/", {
      email,
      password,
    });
    const token = resp.data.token;

    if (token !== undefined) {
      localStorage.setItem("token", token);
      localStorage.setItem("user", email);

      dispatch(
        login({
          userName: email,
        })
      );
    }
  };
};

/* Accion: verifica si usuario esta autenticado*/
export const startChecking = () => {
  return (dispatch) => {
    let user = localStorage.getItem("user");
    dispatch(checking(user));
  };
};

const login = (user) => {
  return {
    type: types.authLogin,
    payload: user,
  };
};

const checking = (user) => {
  return {
    type: types.authStartLogin,
    payload: user,
  };
};

/* Accion: logout de usuario */
export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  };
};

const logout = () => {
  return {
    type: types.authLogout,
  };
};
