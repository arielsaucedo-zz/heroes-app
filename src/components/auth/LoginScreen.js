import React from "react";
import { Formik } from "formik";
import { useDispatch } from "react-redux";
import { startLogin } from "../../actions/auth";

import "./login.css";
import Swal from "sweetalert2";

export const LoginScreen = () => {
  const dispatch = useDispatch();

  return (
    <div className="d-flex justify-content-center">
      <Formik
        initialValues={{ email: "challenge@alkemy.org", password: "react" }}
        validate={(values) => {
          const errors = {};
          if (!values.email) {
            errors.email = "Este campo es requerido";
          } else if (
            !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
          ) {
            errors.email = "Ingrese un email correcto";
          }
          if (!values.password) {
            errors.password = "Este campo es requerido";
          }
          return errors;
        }}
        onSubmit={(values) => {
          return dispatch(startLogin(values.email, values.password)).catch(
            () => {
              Swal.fire({
                icon: "error",
                title: "Error!",
                text: "Usuario o contraseÃ±a incorrectos",
              });
            }
          );
        }}
      >
        {({
          values,
          touched,
          errors,
          handleChange,
          handleSubmit,
          isSubmitting,
        }) => (
          <form className="box" onSubmit={handleSubmit}>
            <h1>Login</h1>
            <p className="text-muted"> Please enter your login and password!</p>
            <input
              data-testid="email"
              autoComplete="off"
              type="email"
              name="email"
              placeholder="Username"
              value={values.email}
              onChange={handleChange}
            />
            {errors.email && touched.email && errors.email}
            <input
              data-testid="password"
              type="password"
              name="password"
              placeholder="Password"
              value={values.password}
              onChange={handleChange}
            />
            {errors.password && touched.password && errors.password}
            <input
              data-testid="button"
              className="btn-login"
              type="submit"
              name=""
              value="Login"
              disabled={isSubmitting}
            />
          </form>
        )}
      </Formik>
    </div>
  );
};
