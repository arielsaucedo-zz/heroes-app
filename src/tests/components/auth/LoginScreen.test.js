import { mount } from "enzyme";
import { Provider } from "react-redux";
import "@testing-library/jest-dom/extend-expect";
import { fireEvent, render, waitFor } from "@testing-library/react";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Swal from "sweetalert2";

jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

import { LoginScreen } from "../../../components/auth/LoginScreen";

const initialState = {
  auth: {
    logged: false,
    userName: "",
  },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
);

describe("Pruebas en componente <LoginScreen />", () => {
  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test("Debe de mostrarse correctamente ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe validar campos del login al realizar el submit", async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <LoginScreen />
      </Provider>
    );

    const emailField = getByTestId("email");

    await waitFor(() => {
      fireEvent.change(emailField, { target: { value: "pepe@gmail.com" } });
    });

    const button = getByTestId("button");

    await waitFor(() => {
      fireEvent.click(button);
    });

    await waitFor(() => {
      expect(Swal.fire).toHaveBeenCalled();
    });
  });
});
