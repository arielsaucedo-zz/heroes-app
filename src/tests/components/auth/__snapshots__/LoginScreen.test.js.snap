// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <LoginScreen /> Debe de mostrarse correctamente  1`] = `
<div
  className="d-flex justify-content-center"
>
  <form
    className="box"
    onSubmit={[Function]}
  >
    <h1>
      Login
    </h1>
    <p
      className="text-muted"
    >
       Please enter your login and password!
    </p>
    <input
      autoComplete="off"
      data-testid="email"
      name="email"
      onChange={[Function]}
      placeholder="Username"
      type="email"
      value="challenge@alkemy.org"
    />
    <input
      data-testid="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="react"
    />
    <input
      className="btn-login"
      data-testid="button"
      disabled={false}
      name=""
      type="submit"
      value="Login"
    />
  </form>
</div>
`;
